model_analyzer/result/result_manager.py

```diff
@@ -229,39 +227,56 @@ def add_measurement(self, run_config, measurement):
                 "Cannot add measurements without tables")
-        elif not self._result_comparator:
-            raise TritonModelAnalyzerException(
-                "Cannot add measurements without setting result comparator")
 
         model_name = run_config.model_name()
-        model_config_name = run_config.model_config().get_field('name')
-        if model_config_name not in self._results[model_name]:
-            self._results[model_name][model_config_name] = ModelResult(
-                model_name=run_config.model_name(),
-                model_config=run_config.model_config(),
-                comparator=self._result_comparator,
-                constraints=self._constraints)
-        measurement.set_result_comparator(comparator=self._result_comparator)
-        self._results[model_name][model_config_name].add_measurement(
-            measurement)
-
-    def sort_results(self, model_name):
-        """
-        Takes the results constructed by adding 
-        measurements and puts them in ResultHeaps
-        to be sorted.
-
-        Parameters
-        ----------
-        model_name: str
-            The name of the model whose results are being sorted
-
-        """
-
-        # Add results to individual result heaps as well as global result heap
-        for result in self._results[model_name].values():
-            self._per_model_sorted_results[model_name].add_result(result)
-            self._across_model_sorted_results.add_result(result)
+        model_config = run_config.model_config()
+        model_config_name = model_config.get_field('name')
+
+        # Get reference to results state and modify it
+        results = self._state_manager.get_state_variable(
+            'ResultManager.results')
+
+        if model_name not in results:
+            results[model_name] = {}
+        if model_config_name not in results[model_name]:
+            results[model_name][model_config_name] = (model_config, {})
+
+        measurement_key = measurement.perf_config().to_cli_string()
+        results[model_name][model_config_name][1][
+            measurement_key] = measurement
+
+    def collect_and_sort_results(self):
+        """
+        Collects objectives and constraints for
+        each model, constructs results from the
+        measurements obtained, and sorts and 
+        filters them according to constraints
+        and objectives.
+        """
+
+        # Collect objectives and constraints
+        comparators = {}
+        constraints = {}
+        for model in self._config.model_names:
+            comparators[model.model_name()] = ResultComparator(
+                metric_objectives=model.objectives())
+            constraints[model.model_name()] = model.constraints()
+
+        # Construct and add results to individual result heaps as well as global result heap
+        results = self._state_manager.get_state_variable(
+            'ResultManager.results')
+        for model_name, result_dict in results.items():
```

Comment:
Just trying to understand, in the new version, the results will not be added to the result manager until the very end right?

Reply:
Yes, only measurements are added during profiling, and the results are then constructed and sorted after all the runs have finished.